Jeremy master

JeremS opened the pull request:
Hi,

Here is my take on porting functionalities from the reactrails example to Clojure & Clojurescript.

In the backend we have: 
- [Datomic](http://www.datomic.com/) as a database layer
- [Liberator](http://clojure-liberator.github.io/liberator/) to create REST resources
- [Prismatic's schema](https://github.com/Prismatic/schema) to validate and coerce data
- [bidi](https://github.com/juxt/bidi) for the routing part
- [Immutant](http://immutant.org/) web as webserver
- [Stuart Sierra's component](https://github.com/stuartsierra/component) to try to sanely glue this together.

For the Front end:
- [Reagent](https://github.com/reagent-project/reagent) to interface with react
- [core.async](https://github.com/clojure/core.async) to handle the different of the app in a way close to the one [Petrol](https://github.com/krisajenkins/petrol) provides
- [cljs.ajax](https://github.com/JulianBirch/cljs-ajax) to talk to the backend
- [Prismatic's schema](https://github.com/Prismatic/schema) & [bidi](https://github.com/juxt/bidi) again, same purposes

I have used [figwheel](https://github.com/bhauman/lein-figwheel) to get hot reload while coding. I however didn't use the leiningen plugin. Instead I used figwheel-sidecar directly and a tiny bit of plumbing(script/repl.clj) to make it work with [Cursive](https://cursive-ide.com/).

The app is deployed on [heroku](https://rinr.herokuapp.com/).
To start it locally:
- start the webserver in one repl with `lein repl` then:

``` clojure
(in-ns 'user)
=>...
(go)
...
```
- Start figwheel in another repl with `lein repl` then:
  :

``` clojure
(in-ns 'user)
=>...
(use 'figwheel-sidecar.repl-api)
=>...
(start-figwheel!)
=>... 
(cljs-repl)
=>...
```

Tell me what you think!

Cheers,

Jeremy.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/shakacode/reactrails-in-reagent/5)

<!-- Reviewable:end -->

JeremS committed:
First commit

Resources GET method should work.
Still needs to implement POST/PUT methods, validation and coercion with
schema.

JeremS committed:
Posting comments is possible

JeremS committed:
Support for the static ressources

JeremS committed:
Merge remote-tracking branch 'refs/remotes/origin/datomic-backend'

JeremS committed:
The client can post comments

JeremS committed:
Added simple refresh of the comments

JeremS committed:
Merge branch 'refs/heads/Clojurescript-front-end'

JeremS committed:
Removed petrol dependency

I rolled my own dispatcher, greatly inspired from petrol though.

JeremS committed:
Externalized routes data.

Theses changes allow the server and the client to use the same routes
definition. This way we can use bidi's path-for function inside of both.

JeremS committed:
cleanup

JeremS committed:
Errata correction

JeremS committed:
Refactoring

- The app's handler building process is not statefull anymore
- there is now a separate namespace for building a dev system
- the dev system behaves better with the ring reload middleware

JeremS committed:
minor corrections

JeremS committed:
some cleanup

JeremS committed:
Added some tests

JeremS committed:
More tests & correction

corrected the get comment requests so that we're sure to pull on an
actual comment
and not any exsiting entity

JeremS committed:
Added proper exists? decision for the comment-entry resource

JeremS committed:
Removed passing the app's routes to handlers via the request map.

All the routes are defined in the same namespace and are static,
constant for the lifecycle of the app. No point in passing them via the
request map to the handlers.

JeremS committed:
More assertions

JeremS committed:
Testing the comment-list ressource GET method

JeremS committed:
_

JeremS committed:
correction

JeremS committed:
cleanup

JeremS committed:
Fix weird mistake I made

JeremS committed:
Adding bootstrap

JeremS committed:
Use of bootstrap

JeremS committed:
Rework of the comment view

JeremS committed:
fixed concurency issue

JeremS committed:
Use of a sorted set to merge new comments

JeremS committed:
_

JeremS committed:
Markdown support

JeremS committed:
Refactor of the client

Use of common schemas and ajax.cljs

JeremS committed:
Remove unused namespaces

JeremS committed:
_

JeremS committed:
cleanup

JeremS committed:
_

JeremS committed:
cleanup

JeremS committed:
cleanup

JeremS committed:
Preparing prod version to deploy on heroku

JeremS committed:
Still working on heroku

Not deployable yet

JeremS committed:
Still preparing heroku deployment

JeremS committed:
integer parsing of $PORT

JeremS committed:
Fix

Adding the host there should bind the port provided by heroku and stop
it from killing the app after 60 seconds.

JeremS committed:
Make the host an environment variable

JeremS committed:
Very little polish

justin808 commented on .gitignore:
```diff
@@ -0,0 +1,13 @@
+.idea
+/resources/public/js/compiled
+/resources/public/js/main.js
+.env
```
red marks indicated we're missing a newline at the end of the files

justin808 commented on resources/public/js/marked/0.3.4/marked.min.js:
```diff
@@ -0,0 +1,6 @@
+ * Copyright (c) 2011-2014, Christopher Jeffrey. (MIT Licensed)
+ * https://github.com/chjj/marked
+ */
+(function(){var block={newline:/^\n+/,code:/^( {4}[^\n]+\n*)+/,fences:noop,hr:/^( *[-*_]){3,} *(?:\n+|$)/,heading:/^ *(#{1,6}) *([^\n]+?) *#* *(?:\n+|$)/,nptable:noop,lheading:/^([^\n]+)\n *(=|-){2,} *(?:\n+|$)/,blockquote:/^( *>[^\n]+(\n(?!def)[^\n]+)*\n*)+/,list:/^( *)(bull) [\s\S]+?(?:hr|def|\n{2,}(?! )(?!\1bull )\n*|\s*$)/,html:/^ *(?:comment *(?:\n|\s*$)|closed *(?:\n{2,}|\s*$)|closing *(?:\ ...
```
We want to get this from npm and not put into the repository.

justin808 commented on reactrails-in-reagent.iml:
```diff
@@ -0,0 +1,148 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Not sure if Idea files should go into github. We had issues with ruby projects, so we gitignore all idea files.

justin808 commented on resources/public/html/test.html:
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
```
What's this for?

justin808 commented:
Looks great. Nice to see:
- Tests
- Actually deployed to Heroku

JeremS committed:
Added mention in code.

JeremS committed:
Typo correction

JeremS committed:
Removed unused code

JeremS committed:
Removing IntelliJ iml file from the repo

JeremS committed:
Removed unused file

JeremS committed:
using npm to install front end libraries

JeremS committed:
_

JeremS committed:
Merge remote-tracking branch 'refs/remotes/origin/jeremy-master'

JeremS committed:
correction

JeremS committed:
Merge branch 'jeremy-master'

JeremS committed:
Removing IintelliJ file

JeremS commented:
I misused git at some point, so some of the last commits may not make total sense. The gist of the last modifications is:
- Removal of some unused code like the test.html page
- Removal of the IntelliJ config file from the repo
- Removed marked.js from the repo
- Added some npm scripting to install [marked.js](https://github.com/chjj/marked) automatically

It is worth mentioning that there are multiple ways to include JavaScript libraries into a ClojureScript project. One is to use [CLJSJS](http://cljsjs.github.io/) when the library is packaged there. I am not sure but I believe that JS Libraries packaged there can be minified with the google closure compiler used by
ClojureScript. If that's the case [CLJSJS](http://cljsjs.github.io/) should be the de facto solution when possible in my opinion.

I however didn't use [CLJSJS](http://cljsjs.github.io/) even if I could have. Since all js libraries are not
packaged with [CLJSJS](http://cljsjs.github.io/), I wanted to see what is possible then.
The solution here is to use npm to install [marked.js](https://github.com/chjj/marked) and a npm script to copy the minified version from the `node_modules` directory to the `resources` directory used by lein uberjar. The last piece of the puzzle when it comes to deploy on heroku is to use the heroku/nodejs buildpack before the heroku/clojure buildpack. 
The [current version of the app](https://rinr.herokuapp.com/) running on heroku is using this method and so I was able to remove [marked.js](https://github.com/chjj/marked) from the repo.

Cheers,

Jeremy.

justin808 commented on test/reactrails_in_reagent/system_test.clj:
```diff
@@ -0,0 +1,81 @@
+  (swap! ref component/stop-system))
+
+
+
```
Suggest seeing if there's a Clojure linter to catch things like extra lines.

JeremS replied:
There are some tools that I'll investigate:
- [kibit](https://github.com/jonase/kibit)
- [eastwood](https://github.com/jonase/eastwood)
- [lein-bikeshed](https://github.com/dakrone/lein-bikeshed)

justin808 commented:
@JeremS Looks much cleaner in terms of the PR.

JeremS commented:
Thanks @justin808 
I am also taking a look at [gorilla repl](http://gorilla-repl.org/index.html) and [devcards](https://github.com/bhauman/devcards). It might be the way to make an interactive documentation of the code.

JeremS committed:
Clean up and usage of more centralised schemas

JeremS committed:
First worksheet about the use of schema.

JeremS committed:
Merge branch 'gorilla'

JeremS committed:
Merge branch 'master' into jeremy-master

JeremS committed:
Moving backend tests in their own directories.

JeremS committed:
Missing 404 header

JeremS committed:
Integrating devcards

Devcards seems to work and can run cljs.test!

JeremS committed:
Merge branch 'master' into jeremy-master

JeremS committed:
Added default values for the web server env config.

JeremS committed:
dev system config as a function instead of a static value.

JeremS committed:
devcards!!!

- added devcards
- changed the way to start devcards

JeremS committed:
_

JeremS committed:
_

JeremS committed:
Added Karma for tests

JeremS committed:
Added doc & system rework

The readme now has a section about running the app.
The component have been reworked to share as much as possible between the prod system and the dev one.

JeremS commented:
Time for an update!

## What's new

Testing of the fronted has been added using [Karma](https://karma-runner.github.io/0.13/index.html) as a test runner.

[Gorilla REPL](http://gorilla-repl.org) worksheets can be used to illustrate the backend code. Right now there is one about the use of schema. In the same spirit, for the frontend, there is a new [devcards](https://github.com/bhauman/devcards) config. 

There is a new section in the readme explaining the different ways the start the app and all the tooling used.

## What's next
- I haven't been successful yet with the Clojure linters I have found. I'll work more on that front.
- Refactoring of the `reactrails-in-reagent.dispatch` namespace of the client to make it more testable and amenable to devcards.
- Addition of new devcards and new gorilla worksheets to explain the code.

JeremS committed:
Clean up of the routes defition.

JeremS committed:
Some cleanup.

JeremS committed:
_

justin808 commented:
Thanks. Keep me updated!

JeremS committed:
Correcting omission

JeremS committed:
Use of keywords instead of symbols for naming routes & some cleanup

JeremS committed:
New gorilla worksheet about ring,bidi and liberator

JeremS committed:
Cleanup

JeremS committed:
New gorilla worksheet on the use of le component library.

JeremS committed:
Forgotten corrections

JeremS committed:
Refactor of the dispatcher and added tests.

JeremS committed:
Corrections

JeremS committed:
Correction

The tests weren't properly placed in theirs `testing` envs.

JeremS committed:
After Bikeshed

Removed trailing lines at the end of files and split lines longer than 80 characters pointed by bike shed.

JeremS committed:
finished applying bikeshed recommandations

JeremS committed:
Added a script to use the eastwood linter.

JeremS commented:
New update!

There's 2 new Gorilla worksheets explaining different parts of the backend, the `dispatch` namespace use in the client has been slightly modified to be testable, and tests have been added. On the linter front I have tested: 
- [kibit](https://github.com/jonase/kibit) that found some some threading macros with only one function call 
- [lein bikeshed](https://github.com/dakrone/lein-bikeshed) that found files with trailing lines at the end of files, lines longer than 80 characters
- [lein eastwood](https://github.com/jonase/eastwood) that didn't find anything to say...
